### DISCUSSION TITLE
How max_windows affect dataset generation?

### DISCUSSION THREAD

**Discussion by ammarsani-my** [azavea/raster-vision#1520] [Q&A] [answered]
Hi @AdeelH @lewfish.

In our pipeline, we tried few number for `max_windows`. As of now, the max we obtained is 32. If we set it lower, training and validation dataset windows generated is following that number. But not the case if we set higher than that.

My questions are:
1. How to know the maximum windows we could use for each scene?
2. Why RV set max to 32?
3. If we have multiple datasets/scenes to train later on, what will be the value for this? Is it will randomly picked 32 windows from all those datasets?

Thanks!

**Comment by lewfish** [comment, marked as answer]
The `max_windows` argument is used when `method` is set to `random` as described in these [docs](https://docs.rastervision.io/en/latest/api_reference/_generated/rastervision.pytorch_learner.learner_config.GeoDataWindowConfig.html#rastervision.pytorch_learner.learner_config.GeoDataWindowConfig.max_windows). This controls the maximum number of randomly sampled windows *per scene*. This sets the maximum number of windows as opposed to an exact number of windows since a random sampler is used to get each window with a maximum number of tries (set by `max_sample_attempts`), and the sampler may fail to find a window. This may be surprising, but if the AOI is very small and/or has a complex shape, it may be difficult to randomly sample windows that fit within that AOI. The default value is 10,000. I'm not sure where you are getting the value of 32 from. Could you point me to the source?

**Reply by ammarsani-my** [reply]
Thanks for the reply.

I have re-tried to unset the `max_windows`:
```
window_opts = GeoDataWindowConfig(
    method=GeoDataWindowMethod.random,
    size=chip_sz,
    size_lims=(chip_sz, chip_sz + 1),
)
```

Then re-run the pipeline. These logs were printed:
`2022-11-01 06:25:15:rastervision.pytorch_learner.learner: INFO - train_ds: 10000 items`
`2022-11-01 06:25:15:rastervision.pytorch_learner.learner: INFO - valid_ds: 10000 items`

However, seeing at `dataloaders/train.png` and `dataloaders/valid.png`, the windows generated displayed there is still 32 windows.

**Reply by ammarsani-my** [reply]
One more related question. May I know what is the 313 stands for?
```
Training: 100%|█████████████████████████████████| 313/313 [02:43<00:00,  1.91it/s]
```
Thanks!

**Reply by lewfish** [reply]
The `train.png` and `valid.png` files just show a sample of the windows for debugging purposes. I think the number of samples is equal to the `batch_sz`. The 313 is the number of batches per epoch which is roughly equal to the total number of samples divided by the batch size. (An epoch encompasses a single run through all the samples.)

**Reply by ammarsani-my** [reply]
That's a great crystal clear explanation@lewfish! I'll see my configuration shortly to understand better on this.

**Reply by ammarsani-my** [reply]
Hi @lewfish. I just reviewed the config.

> I think the number of samples is equal to the batch_sz.

For now, our `batch_sz` was remain as the default, which is 32. If number of samples was the same means there is 32 samples. What would be the calculation for total number of samples then to be divided with the batch size?

@AdeelH if you would like to add in something, I would appreciate it so much. 

Appreciate all your help on this.
Once again, thanks!

**Reply by lewfish** [reply]
Actually, I wasn't totally clear :) I meant that the number of windows shown in `train.png` and `valid.png` is equal to the `batch_sz`. The total number of samples in the dataset is 10,000. I know that because of the logs you showed above:

```
2022-11-01 06:25:15:rastervision.pytorch_learner.learner: INFO - train_ds: 10000 items
2022-11-01 06:25:15:rastervision.pytorch_learner.learner: INFO - valid_ds: 10000 items
```

So, the (ceiling of the) number of samples divided by the batch size is the number of steps (ie. batches processed) per epoch, which in your case is 313.

**Reply by ammarsani-my** [reply]
Okay @lewfish.
That sounds really make sense for me 👍

Thanks for this explanation.
I'll try to figure it out from here.
